### PR TITLE
Fix Issue #41822 Add IsNullOrEmpty or IsNullOrWhiteSpace check for method parameters throws an empty "message"

### PR DESCRIPTION
**Pull request by Sliptory**
#41822

### CODE REVIEW

File: src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs [hunk 1]
```diff
@@ -35,6 +35,8 @@ internal abstract partial class AbstractAddParameterCheckCodeRefactoringProvider
     {
         protected abstract bool CanOffer(SyntaxNode body);
         protected abstract bool PrefersThrowExpression(DocumentOptionSet options);
+        protected abstract SyntaxToken CreateInterpolatedStringStartToken();
```
Comment (YairHalberstadt) [inline src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs]: Maybe move these to SyntaxGenerator, and keep everything here static?
Then you could also remove the abstract methods from AbstractConvertConcatenationToInterpolatedStringRefactoringProvider.cs which seems to be the other user of this generator API.

File: src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs [hunk 2]
```diff
@@ -1428,7 +1428,7 @@ public C(string s)
     {
         if (string.IsNullOrWhiteSpace(s))
         {
-            throw new ArgumentException(""message"", nameof(s));
+            throw new ArgumentException($""'{nameof(s)}' cannot be null or empty"", nameof(s));
```
Comment (YairHalberstadt) [inline src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs]: Shouldn't this be cannot be null or whitespace?

File: src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs [hunk 3]
```diff
@@ -518,22 +520,30 @@ private static TStatementSyntax CreateNullCheckStatement(SemanticModel semanticM
...
-        private static SyntaxNode CreateArgumentException(
+        private SyntaxNode CreateArgumentException(
             Compilation compilation, SyntaxGenerator generator, IParameterSymbol parameter)
         {
-            // Note "message" is not localized.  It is the name of the first parameter of 
-            // "ArgumentException"
+            var content = new List<SyntaxNode>()
+            {
+                generator.InterpolatedStringText(generator.InterpolatedStringTextToken($"'{{{generator.NameOfExpression(generator.IdentifierName(parameter.Name))}}}'")),
+                generator.InterpolatedStringText(generator.InterpolatedStringTextToken("cannot be null or empty"))
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs]: this would need to be localized.

File: src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs [hunk 4]
```diff
@@ -518,22 +520,30 @@ private static TStatementSyntax CreateNullCheckStatement(SemanticModel semanticM
...
+            };
             return generator.ObjectCreationExpression(
                 GetTypeNode(compilation, generator, typeof(ArgumentException)),
-                generator.LiteralExpression("message"),
+                generator.InterpolatedStringExpression
+                (
+                    CreateInterpolatedStringStartToken().WithLeadingTrivia(content.First().GetLeadingTrivia()),
+                    content,
+                    CreateInterpolatedStringEndToken().WithTrailingTrivia(content.Last().GetTrailingTrivia())
+                ),
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs]: non-idiomatic formatting for this layer.

File: src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs [hunk 5]
```diff
@@ -504,7 +506,7 @@ private static TStatementSyntax CreateNullCheckStatement(SemanticModel semanticM
             return null;
         }
 
-        private static SyntaxNode GetTypeNode(
+        private SyntaxNode GetTypeNode(
             Compilation compilation, SyntaxGenerator generator, Type type)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs]: why was this made non-static?

File: src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs (hunk 2 above)
Comment (Sliptory) [reply src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs]: nope

File: src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs (hunk 1 above)
Comment (Sliptory) [reply src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs]: This is good idea!

File: src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs (hunk 2 above)
Comment (YairHalberstadt) [reply src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs]: Why not?

Comment (Sliptory) [reply src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs]: I do not know in what context this was called exception

File: src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs (hunk 4 above)
Comment (Sliptory) [reply src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs]: need to change
`content.First()`
to
`content[0]`
for example?

File: src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs]: i think @YairHalberstadt  is simply saying that the string should be `$""'{nameof(s)}' cannot be null or whitespace""`  since that's the check that is actually happening.

File: src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs]: no, this is about the formatting, we don't break invocations up like this where the open paren and close paren are on different lines.

Comment (Sliptory) [reply src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs]: i understood)

File: src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs (hunk 3 above)
Comment (Sliptory) [reply src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs]: ok

File: src/Workspaces/VisualBasic/Portable/CodeGeneration/VisualBasicSyntaxGenerator.vb [hunk 6]
```diff
@@ -48,6 +48,13 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.CodeGeneration
             Return SyntaxFactory.SeparatedList(Of TElement)(list)
         End Function
 
+        Friend Overrides Function CreateInterpolatedStringStartToken(isVerbatim As Boolean) As SyntaxToken
+            Return SyntaxFactory.Token(SyntaxKind.DollarSignDoubleQuoteToken)
+        End Function
+
+        Friend Overrides Function CreateInterpolatedStringEndToken() As SyntaxToken
+            Return SyntaxFactory.Token(SyntaxKind.DoubleQuoteToken)
```
Comment (JoeRobich) [inline src/Workspaces/VisualBasic/Portable/CodeGeneration/VisualBasicSyntaxGenerator.vb:56]:
```suggestion
            Return SyntaxFactory.Token(SyntaxKind.DoubleQuoteToken)
        End Function
```

File: src/Features/Core/Portable/FeaturesResources.resx [hunk 7]
```diff
@@ -1460,6 +1460,16 @@ This version used in: {2}</value>
   <data name="Implement_explicitly" xml:space="preserve">
     <value>Implement explicitly</value>
   </data>
+  <data name="cannot_be_null_or_empty" xml:space="preserve">
+    <value>cannot be null or empty</value>
```
Comment (JoeRobich) [inline src/Features/Core/Portable/FeaturesResources.resx:1464]: Please add a `{0}` place holder for replacement with the parameter name.  Constructing this string with concatenation may not make sense for all languages.

File: src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs [hunk 8]
```diff
@@ -330,17 +330,17 @@ public C(string a, string b, string c)
     {
         if (string.IsNullOrEmpty(a))
         {
-            throw new ArgumentException(""message"", nameof(a));
+            throw new ArgumentException($""'{nameof(a)}' cannot be null or empty"", nameof(a));
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs:333]: tests will fail on spanish leg if they reference english strings.

File: src/Features/Core/Portable/FeaturesResources.resx [hunk 9]
```diff
@@ -1460,6 +1460,16 @@ This version used in: {2}</value>
...
+  </data>
+  <data name="cannot_be_null_or_whitespace" xml:space="preserve">
+    <value>cannot be null or whitespace</value>
+  </data>
+  <data name="Struct_contains_assignment_to_this_outside_of_constructor_Make_readonly_fields_writable" xml:space="preserve">
+    <value>Struct contains assignment to 'this' outside of constructor. Make readonly fields writable</value>
+    <comment>{Locked="Struct"}{Locked="this"} these are C#/VB keywords and should not be localized.</comment>
+  </data>
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/FeaturesResources.resx:1472]: why was this added?

File: src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs [hunk 10]
```diff
@@ -537,13 +537,25 @@ private static TStatementSyntax CreateNullCheckStatement(SemanticModel semanticM
...
+            var content = new List<SyntaxNode>()
+            {
+                generator.InterpolatedStringText(generator.InterpolatedStringTextToken($"'{{{generator.NameOfExpression(generator.IdentifierName(parameter.Name))}}}'")),
+                generator.InterpolatedStringText(generator.InterpolatedStringTextToken(text))
+            };
             return generator.ObjectCreationExpression(
                 GetTypeNode(compilation, generator, typeof(ArgumentException)),
-                generator.LiteralExpression("message"),
+                generator.InterpolatedStringExpression(
+                    generator.CreateInterpolatedStringStartToken(false).WithLeadingTrivia(content.First().GetLeadingTrivia()),
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs:556]: named arg for `false`

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs:556]: the first/last content nodes won't have any trivia.  so i don't know why w're taking trivia from them and adding to the interpolated string.

File: src/Features/Core/Portable/xlf/FeaturesResources.cs.xlf [hunk 11]
```diff
@@ -2420,6 +2420,26 @@ If the "M" format specifier is used without other custom format specifiers, it's
         <target state="new">The "MMM" custom format specifier represents the abbreviated name of the month. The localized abbreviated name of the month is retrieved from the DateTimeFormatInfo.AbbreviatedMonthNames property of the current or specified culture.</target>
         <note />
       </trans-unit>
+      <trans-unit id="_0_can_be_simplified">
+        <source>{0} can be simplified</source>
+        <target state="translated">{0} se dá zjednodušit.</target>
+        <note />
+      </trans-unit>
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/xlf/FeaturesResources.cs.xlf:2427]: why was this aded?

**Comment by CyrusNajmabadi** [comment]
@Sliptory If you'd like, we can pick up this PR from you and finish it.  Thanks!

**Comment by Sliptory** [comment]
> @Sliptory If you'd like, we can pick up this PR from you and finish it. Thanks!

Ok, Thanks!)

**Comment by CyrusNajmabadi** [comment]
@JoeRobich i can take this unless you wanted to :)

**Comment by JoeRobich** [comment]
> @JoeRobich i can take this unless you wanted to :)

@CyrusNajmabadi Please, be my guest =)

**Comment by CyrusNajmabadi** [comment]
@JoeRobich this is ready for review.  Thanks!

**Comment by CyrusNajmabadi** [comment]
Thanks @Sliptory .  Sorry that took so long :)
